paywall/src/unlock.js/Wallet.ts

```diff
@@ -0,0 +1,315 @@
+  private readonly isMetamask: boolean
+  private readonly config: PaywallConfig
+  private hasWeb3: boolean = false
+  private useUserAccounts: boolean = false
+
+  private userAccountAddress: string | null = null
+  private userAccountNetwork: number
+
+  constructor(
+    window: Web3Window,
+    iframes: IframeHandler,
+    config: PaywallConfig,
+    constants: StartupConstants
+  ) {
+    this.window = window
+    this.iframes = iframes
+    this.config = config
+    this.userAccountNetwork = constants.network
+
+    // do we have a web3 wallet?
+    this.hasWallet = !!(this.window.web3 && this.window.web3.currentProvider)
+    this.isMetamask = !!(
+      window.web3 &&
+      window.web3.currentProvider &&
+      window.web3.currentProvider.isMetamask
+    )
+    // user accounts are used in 2 conditions:
+    // 1. there is no crypto wallet present
+    // 2. the paywall configuration explicitly asks for them
+    this.useUserAccounts =
+      !this.hasWallet &&
+      (config.unlockUserAccounts === true ||
+        config.unlockUserAccounts === 'true')
+    if (process.env.DEBUG) {
+      if (this.useUserAccounts) {
+        // eslint-disable-next-line
+        console.log('[USER ACCOUNTS] using user accounts')
+      } else {
+        // eslint-disable-next-line
+        console.log('[USER ACCOUNTS] using native crypto wallet')
+      }
+    }
+  }
+
+  init() {
+    if (this.useUserAccounts) {
+      // create the preconditions for using user accounts
+      this.setupUserAccounts()
+    }
+    // set up the proxy wallet
+    this.setupProxyWallet()
+  }
+
+  /**
+   * This is called only if we can use user accounts
+   */
+  setupUserAccounts() {
+    // listen for updates to state from the data iframe, and forward them to the checkout UI
+    this.iframes.data.on(PostMessages.UPDATE_LOCKS, locks =>
+      this.iframes.accounts.postMessage(PostMessages.UPDATE_LOCKS, locks)
+    )
+
+    // pass on the configuration and request the latest data
+    this.iframes.accounts.on(PostMessages.READY, () => {
+      this.iframes.accounts.postMessage(PostMessages.CONFIG, this.config)
+
+      this.iframes.data.postMessage(PostMessages.SEND_UPDATES, 'locks')
+    })
+
+    // listen for account and network from the user accounts iframe
+    this.iframes.accounts.on(PostMessages.UPDATE_ACCOUNT, account => {
+      this.userAccountAddress = account
+    })
+    this.iframes.accounts.on(PostMessages.UPDATE_NETWORK, network => {
+      this.userAccountNetwork = network
+    })
+
+    // when a purchase is in progress, tell the data iframe to retrieve the transaction
+    this.iframes.accounts.on(PostMessages.INITIATED_TRANSACTION, () => {
+      this.iframes.data.postMessage(
+        PostMessages.INITIATED_TRANSACTION,
+        undefined
+      )
+    })
+
+    // then create the iframe and ready its post office
+    this.iframes.accounts.createIframe()
+  }
+
+  setupProxyWallet() {
+    if (this.useUserAccounts && !this.hasWallet) {
```

Comment:
Isn't that already checked when calling `setupProxyWallet` above?

Reply:
Maybe we can get rid of that function and move all of its logic inside of the function calling it because it seems to be checking for the same things?

Reply:
ok, I moved setupProxyWallet inside `init()`